src/librustc_mir/interpret/eval_context.rs

```diff
@@ -729,6 +724,17 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         let frame =
             self.stack_mut().pop().expect("tried to pop a stack frame, but there were none");
 
+        if !unwinding {
+            // Copy the return value to the caller's stack frame.
+            if let Some(return_place) = frame.return_place {
+                let op = self.access_local(&frame, mir::RETURN_PLACE, None)?;
+                self.copy_op_transmute(op, return_place)?;
```

Comment:
Not sure if we can pull this off, but maybe we can (for non-immediates) make the return place alias with the frame's `RETURN_PLACE` local. This would be doable by changing the `LocalState` of the `RETURN_PLACE` to `Live(Operand::Indirect(return_place))` (well you still need to convert the https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.PlaceTy.html to an MPlaceTy, but that should be doable (and where not, we have an immediate, which is cheap and we can just keep copying it here)

Reply:
Yeah, I was wondering the same thing. We need to be careful then when popping the frame to not do `copy_op` (as that would overlap itself) but still do validation.

The main subtlety here is making sure this is a Miri-internal optimization that cannot be observed by the program. The fact that we reborrow the return place should exclude any observations by accessing the return place directly... so I think the only remaining possible observation is through address equality tests.

Reply:
So... about this PR. I'd be fine merging it with the 8% regression on the stress tests (with an open issue for fixing it again). Then we can experiment with this local aliasing trick in a PR that does nothing but the optimization.

Reply:
That would be nice. I don't think I have the bandwidth for these deeper changes to const eval.

Reply:
Is that something we should nominate for T-compiler decision?

Reply:
I think @rust-lang/wg-const-eval can make this decision on our own, since it's just a stress test being regressed. Regressing the test is not an issue, we just have the test to make sure we don't do so accidentally

Reply:
Okay, fine for me. @ecstatic-morse what do you think?

@jonas-schievink can you create an issue describing the perf regression and linking to this discussion?

Reply:
Opened https://github.com/rust-lang/rust/issues/71463